Integratin test flake: persistent_volumes_test.go:237: Only 1 PV should be bound but got 0

The flake is encountered after I enable the watch cache in #26494. I don't know if it's related. The flake shouldn't occur either with watch cache disabled or enabled.

```
--- FAIL: TestPersistentVolumeMultiPVs (0.53s)
	persistent_volumes_test.go:237: Only 1 PV should be bound but got 0
```

[logs](https://console.cloud.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/26494/kubernetes-pull-test-unit-integration/28367/)
[.stdout](https://00e9e64bac63eec1b40fd593689906a9a117f39f53fd9e33d1-apidata.googleusercontent.com/download/storage/v1_internal/b/kubernetes-jenkins/o/pr-logs%2Fpull%2F26494%2Fkubernetes-pull-test-unit-integration%2F28367%2Fartifacts%2Fjunit_v1,extensions-v1beta1,metrics-v1alpha1,federation-v1alpha1_20160528-165200.stdout?qk=AD5uMEvMTZAwik1YrkZsvNnBBXSXPC0iWRLj3eGr10qJqoLtCQHuLW4DYnoA9CJsyLyq4Hxj6ZEDTOM2rcpe9EKYk0ZViNNC4q5IBVvzdFlMz_T0EtNkSBk8yHRZyfLnvdkNVp0vBPoKOb3PPkod_74WWPCG64SyygJzPcNtZhTFyF0lwHIGIThSqY5S77QJIGwbInAtC_WQuGFuz-sz_1TdX8VRLy6wht3-Nzshz6HnoB1Bm3q7w5jZElp6LiF7wFba5uollWC-zsGSnC4ouJ_Kc4lfowkSbHUgiAa4N0V-5h2NMVdU4umZaS0hPmlbZGDU_NMps1CQarLCk_cNZnlsrKLkerMg2yHvP65Im2VIjjt0PbQi2F7fZ1WbGbTGf9TrscTCECasmIAuVTFrADfcLGF9NNsy2DOj7IIAJ7aqT5XLCvxch5FmwMJzE3ppRROoQ3bah0mlNsD8pMVsmtOB3V925cwF6SHQJRUV64U89mpFUbDSd0FIkE6CZggiN_5oAT4MiPc78MxJhPgP-LlfpSakqlTH6W3-wVjB1e33HEvlrnxuUBGl2wH-uU-lc56m2eE7kRpzDgJhXcAwawQY1L_LXeuqF1yc1M5RodsAMOTsyf9OkuCKM6q0eYTypF6uEoeY836BXsWA92T-0SZGM4de7ICW0dUeOhr28QMXPZCuKyFI8ksgUYdGucESpb_JNTbOfRZzHbmCR3xIc4iKj17j_8qUNSZpB4ycsAcoGKExDVr0nVDx46BEMk4wpsGK8fMfdehPcgqxSwQLkS-wvkP1j6dSbfA2P7K76_x9jB0m2n-fG8fG5g4BCit1GfiMbgBU4IRWMMAFDgdAsAlR9SyVnvAYomOLMbEC1hoYycIsQGTBYKJF36nrGmgf1uHlOZraK7GHgzWvFLgTyCSonpL8K1fkvxrS6b6yLdrnjzQbpvWE49OELMRWlThlffvQgWdj7FryU6a0Nwg0cJc4KWo3FEW54dHh7UMdmosnTUf6auTyeSQkZlXieZRlX0xHfuWUEw01dafyw37vd2IcUPSFUjSVDw)
[build-log](https://00e9e64bacd3c073c0cb10530fd42ca03b89bebbde001b9252-apidata.googleusercontent.com/download/storage/v1_internal/b/kubernetes-jenkins/o/pr-logs%2Fpull%2F26494%2Fkubernetes-pull-test-unit-integration%2F28367%2Fbuild-log.txt?qk=AD5uMEujCOAYhsFm8MpzR0kHssgUqt4BQkjuE-tOKy1NqC-aNYrZtBU9aMFt5EerV5gvG1a3XEvppcXokEd6wLx65dwzL4INLyoWjx7bNOoQAjQo1vYbjv93UpBH0b6O6ost_EnaJ0quum07yx6P7LOoX4JBOEd5fa1a2e6PQeODl772dxJPsI1Uu9WxrsJ1IZf0fZ6SM-aVf03yAUgZkpF5XMpIfvZoSVu4JxV60NxAwPbPEXKyZlFjIhNj7N3rOwmdCtXHcQ73fuer-HtOlHqLr8YdqNOE1utahRWcz9JxwiQr12ci1r3BArv5XbIBmMLKGsIrSncbpahgBo30xXIBkdYG9hG1nd3qLtKSY_IU5b_qgfzjKrxNJc4mp4DriA_-qw3JFAGt41-Z2r1LwFvcZT4HJ5i-oguSpr_YwGWOtdrTzcEy20btPky3bqDDTIHVUMZE7KpdJ5XfEQKbXQMBMLiJ0OPzOipXKAGwzH0_qZJV-cQhbYvQLHnisk1MWzIjTS4h8wt7wRat2IeUEIXLM-6ByAzr2OhVzIqo-dEvo8Qvqvl5MPyuuOqGbeopEsDgQ3WeRCHeUHJZtJLQNubPOOTlTo9_5xjNYeYsj9TSDgmGyjr8w6V2rY8eB3mgYP3rD-QWSYKFp5IH5kiwta5rjgzutB_LAGa2BEdt2Ynzwz73L1gilG2YQmZxF6Wy5f_e81meYCU_AVPzb2TiMxODdM1t2uyTcAanLRTmygsC0GfPayhWkQcEvyF4c71bN5mNhJrX84QB1T2jpNxEaWVdTANF0OJPkEYoCKcSnWbbxcn0emf-t3DGDgh7CMV0BLL0QMqScBVCZuof3Utthe-6fKfi-Eu26CNG2Siz9y6T-pjkG_D1bFWxmUnAwPJ1gY9wUdnIPRmMQ-HbGBOlPHNTej8GeKIC-ngqf-5VhHxhMYwPoAJBhs8)

@saad-ali

[latest .stdout](https://00e9e64bac65dc7ada0028a5ca5e3bab2965a1402914f2bbc4-apidata.googleusercontent.com/download/storage/v1_internal/b/kubernetes-jenkins/o/pr-logs%2Fpull%2F26494%2Fkubernetes-pull-test-unit-integration%2F28438%2Fartifacts%2Fjunit_v1,extensions-v1beta1_20160529-204714.stdout?qk=AD5uMEs4qoZb2h3CRlSTAmM96d0rt8XB6X02iMxg9IBRebNBzIyyAJaHLAqtrjsA65beNrF79rSj5UWg84882zjacs3YxMgF4na5JiIuL1OCOdO7o8zuU6Q0J74smmdzQ1VgU0hTwO28NFHf97PE1lb-VJUl2Ilccf-D5NKwwmLliFpsS9yUEhb6gR6r3CbLdWmcnFA-Q7ahacS_BQPOO8JSxwuFIN-NlKri9XZ5-nouHDHOQqtbWi1rNHNCpOTtMlzzjNYHzQvXMujMB3l0nk2NgvI0mngMf7brJHb8nZ0ZEzjZX0GWqUCpfQYPDGouIELkU8lzDvCkL2NBmwCQ-21jdET_9ldrabPGiy-pW8vxn6AQ5nueLIEAEQZUqyFv2vzsP-tTxnB9J9QqJUh9pfx4LnwmomxnNzdRKcwpQshYtXubY6M-hwr2he8TJS2HKupJPgFia1UMiRkBRXX6lQfDBQSYrPTdYncZbeNS2cv2xin1UFpiNqm0OWbnoElhbgAo7BxcYYhjLUikSixb0cSFIkIt5QngBhpemt7a90wYguYvF-Mnbl-Hwjx5s5_D08rZJczIkYlYrif50XwsAEONn4A_oJVnq7bfzvuQ7IbLdWCykiD6RgsSyhFHPs_lIdrlZZV2q7i5Pmm0jxOi36J-sKYHgRklwo1nZeNgx0LzEAr3iRu0Bczou-6pQ2TYl4TSYn9zH6FKpbpyi-FNMSh25XX_1yN3ZjSsf5W-vpDZzkYJ1OjIKGPqqJFu_sgcdlSOwtitbLb3blVogFstSBedSc2qOPl24VHmLBVasAQnSzXrs9AiRhJS8IsXlRqZPHQH_gnmfiw_Gy7dT372WmhAK-BkLOPbf3N4FJj7fTVFAJ1UjVSpAfsdV_U0Fl7AnfZtCRywoqpRTdvzUn-y1ksW9jgoXCwKYqzkQAE_swrAUYt-uKef_ixvnaAfwN7863IMZZ7jFipkbKLC4cT2MYXytXJPGy_xPaL6BsEUGnSUjhYkQzWwXEU)

related failure message:
```
--- FAIL: TestPersistentVolumeMultiPVs (0.51s)
	persistent_volumes_test.go:237: Only 1 PV should be bound but got 0
```

All corresponding [logs](https://console.cloud.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/26494/kubernetes-pull-test-unit-integration/28438/).

--- a/test/integration/persistent_volumes_test.go
+++ b/test/integration/persistent_volumes_test.go
@@ -49,7 +49,9 @@ func TestPersistentVolumeRecycler(t *testing.T) {
 	defer s.Close()
 
 	deleteAllEtcdKeys()
-	testClient, ctrl := createClients(s)
+	testClient, ctrl, watchPV, watchPVC := createClients(t, s)
+	defer watchPV.Stop()
+	defer watchPVC.Stop()
 
 	ctrl.Run()
 	defer ctrl.Stop()
@@ -59,13 +61,7 @@ func TestPersistentVolumeRecycler(t *testing.T) {
 
 	pvc := createPVC("fake-pvc", "5G", []api.PersistentVolumeAccessMode{api.ReadWriteOnce})
 
-	w, err := testClient.PersistentVolumes().Watch(api.ListOptions{})
-	if err != nil {
-		t.Errorf("Failed to watch PersistentVolumes: %v", err)
-	}
-	defer w.Stop()
-
-	_, err = testClient.PersistentVolumes().Create(pv)
+	_, err := testClient.PersistentVolumes().Create(pv)
 	if err != nil {
 		t.Errorf("Failed to create PersistentVolume: %v", err)
 	}
@@ -76,15 +72,16 @@ func TestPersistentVolumeRecycler(t *testing.T) {
 	}
 
 	// wait until the controller pairs the volume and claim
-	waitForPersistentVolumePhase(w, api.VolumeBound)
+	waitForPersistentVolumePhase(watchPV, api.VolumeBound)
+	waitForPersistentVolumeClaimPhase(watchPVC, api.ClaimBound)
 
 	// deleting a claim releases the volume, after which it can be recycled
 	if err := testClient.PersistentVolumeClaims(api.NamespaceDefault).Delete(pvc.Name, nil); err != nil {
 		t.Errorf("error deleting claim %s", pvc.Name)
 	}
 
-	waitForPersistentVolumePhase(w, api.VolumeReleased)
-	waitForPersistentVolumePhase(w, api.VolumeAvailable)
+	waitForPersistentVolumePhase(watchPV, api.VolumeReleased)
+	waitForPersistentVolumePhase(watchPV, api.VolumeAvailable)
 
 	// end of Recycler test.
 	// Deleter test begins now.
@@ -95,12 +92,6 @@ func TestPersistentVolumeRecycler(t *testing.T) {
 	// change the reclamation policy of the PV for the next test
 	pv.Spec.PersistentVolumeReclaimPolicy = api.PersistentVolumeReclaimDelete
 
-	w, err = testClient.PersistentVolumes().Watch(api.ListOptions{})
-	if err != nil {
-		t.Errorf("Failed to watch PersistentVolumes: %v", err)
-	}
-	defer w.Stop()
-
 	_, err = testClient.PersistentVolumes().Create(pv)
 	if err != nil {
 		t.Errorf("Failed to create PersistentVolume: %v", err)
@@ -110,17 +101,18 @@ func TestPersistentVolumeRecycler(t *testing.T) {
 		t.Errorf("Failed to create PersistentVolumeClaim: %v", err)
 	}
 
-	waitForPersistentVolumePhase(w, api.VolumeBound)
+	waitForPersistentVolumePhase(watchPV, api.VolumeBound)
+	waitForPersistentVolumeClaimPhase(watchPVC, api.ClaimBound)
 
 	// deleting a claim releases the volume, after which it can be recycled
 	if err := testClient.PersistentVolumeClaims(api.NamespaceDefault).Delete(pvc.Name, nil); err != nil {
 		t.Errorf("error deleting claim %s", pvc.Name)
 	}
 
-	waitForPersistentVolumePhase(w, api.VolumeReleased)
+	waitForPersistentVolumePhase(watchPV, api.VolumeReleased)
 
 	for {
-		event := <-w.ResultChan()
+		event := <-watchPV.ResultChan()
 		if event.Type == watch.Deleted {
 			break
 		}
@@ -157,7 +149,8 @@ func TestPersistentVolumeRecycler(t *testing.T) {
 		t.Fatalf("Unexpected error creating pv: %v", err)
 	}
 
-	waitForPersistentVolumePhase(w, api.VolumeBound)
+	waitForPersistentVolumePhase(watchPV, api.VolumeBound)
+	waitForPersistentVolumeClaimPhase(watchPVC, api.ClaimBound)
 
 	pv, err = testClient.PersistentVolumes().Get(pv.Name)
 	if err != nil {
@@ -176,7 +169,10 @@ func TestPersistentVolumeMultiPVs(t *testing.T) {
 	defer s.Close()
 
 	deleteAllEtcdKeys()
-	testClient, controller := createClients(s)
+	testClient, controller, watchPV, watchPVC := createClients(t, s)
+	defer watchPV.Stop()
+	defer watchPVC.Stop()
+
 	controller.Run()
 	defer controller.Stop()
 
@@ -185,31 +181,30 @@ func TestPersistentVolumeMultiPVs(t *testing.T) {
 	for i := 0; i < maxPVs; i++ {
 		// This PV will be claimed, released, and deleted
 		pvs[i] = createPV("pv-"+strconv.Itoa(i), "/tmp/foo"+strconv.Itoa(i), strconv.Itoa(i)+"G",
-			[]api.PersistentVolumeAccessMode{api.ReadWriteOnce}, api.PersistentVolumeReclaimDelete)
+			[]api.PersistentVolumeAccessMode{api.ReadWriteOnce}, api.PersistentVolumeReclaimRetain)
 	}
 
 	pvc := createPVC("pvc-2", strconv.Itoa(maxPVs/2)+"G", []api.PersistentVolumeAccessMode{api.ReadWriteOnce})
 
-	w, err := testClient.PersistentVolumes().Watch(api.ListOptions{})
-	if err != nil {
-		t.Errorf("Failed to watch PersistentVolumes: %v", err)
-	}
-	defer w.Stop()
-
 	for i := 0; i < maxPVs; i++ {
-		_, err = testClient.PersistentVolumes().Create(pvs[i])
+		_, err := testClient.PersistentVolumes().Create(pvs[i])
 		if err != nil {
 			t.Errorf("Failed to create PersistentVolume %d: %v", i, err)
 		}
 	}
+	t.Log("volumes created")
 
-	_, err = testClient.PersistentVolumeClaims(api.NamespaceDefault).Create(pvc)
+	_, err := testClient.PersistentVolumeClaims(api.NamespaceDefault).Create(pvc)
 	if err != nil {
 		t.Errorf("Failed to create PersistentVolumeClaim: %v", err)
 	}
+	t.Log("claim created")
 
 	// wait until the controller pairs the volume and claim
-	waitForPersistentVolumePhase(w, api.VolumeBound)
+	waitForPersistentVolumePhase(watchPV, api.VolumeBound)
+	t.Log("volume bound")
+	waitForPersistentVolumeClaimPhase(watchPVC, api.ClaimBound)
+	t.Log("claim bound")
 
 	// only one PV is bound
 	bound := 0
@@ -232,6 +227,7 @@ func TestPersistentVolumeMultiPVs(t *testing.T) {
 		t.Logf("claim bounded to %s capacity %v", pv.Name, pv.Spec.Capacity[api.ResourceStorage])
 		bound += 1
 	}
+	t.Log("volumes checked")
 
 	if bound != 1 {
 		t.Fatalf("Only 1 PV should be bound but got %d", bound)
@@ -241,8 +237,10 @@ func TestPersistentVolumeMultiPVs(t *testing.T) {
 	if err := testClient.PersistentVolumeClaims(api.NamespaceDefault).Delete(pvc.Name, nil); err != nil {
 		t.Errorf("error deleting claim %s", pvc.Name)
 	}
+	t.Log("claim deleted")
 
-	waitForPersistentVolumePhase(w, api.VolumeReleased)
+	waitForPersistentVolumePhase(watchPV, api.VolumeReleased)
+	t.Log("volumes released")
 
 	deleteAllEtcdKeys()
 }
@@ -252,25 +250,22 @@ func TestPersistentVolumeMultiPVsDiffAccessModes(t *testing.T) {
 	defer s.Close()
 
 	deleteAllEtcdKeys()
-	testClient, controller := createClients(s)
+	testClient, controller, watchPV, watchPVC := createClients(t, s)
+	defer watchPV.Stop()
+	defer watchPVC.Stop()
+
 	controller.Run()
 	defer controller.Stop()
 
 	// This PV will be claimed, released, and deleted
 	pv_rwo := createPV("pv-rwo", "/tmp/foo", "10G",
-		[]api.PersistentVolumeAccessMode{api.ReadWriteOnce}, api.PersistentVolumeReclaimDelete)
+		[]api.PersistentVolumeAccessMode{api.ReadWriteOnce}, api.PersistentVolumeReclaimRetain)
 	pv_rwm := createPV("pv-rwm", "/tmp/bar", "10G",
-		[]api.PersistentVolumeAccessMode{api.ReadWriteMany}, api.PersistentVolumeReclaimDelete)
+		[]api.PersistentVolumeAccessMode{api.ReadWriteMany}, api.PersistentVolumeReclaimRetain)
 
 	pvc := createPVC("pvc-rwm", "5G", []api.PersistentVolumeAccessMode{api.ReadWriteMany})
 
-	w, err := testClient.PersistentVolumes().Watch(api.ListOptions{})
-	if err != nil {
-		t.Errorf("Failed to watch PersistentVolumes: %v", err)
-	}
-	defer w.Stop()
-
-	_, err = testClient.PersistentVolumes().Create(pv_rwm)
+	_, err := testClient.PersistentVolumes().Create(pv_rwm)
 	if err != nil {
 		t.Errorf("Failed to create PersistentVolume: %v", err)
 	}
@@ -278,14 +273,19 @@ func TestPersistentVolumeMultiPVsDiffAccessModes(t *testing.T) {
 	if err != nil {
 		t.Errorf("Failed to create PersistentVolume: %v", err)
 	}
+	t.Log("volumes created")
 
 	_, err = testClient.PersistentVolumeClaims(api.NamespaceDefault).Create(pvc)
 	if err != nil {
 		t.Errorf("Failed to create PersistentVolumeClaim: %v", err)
 	}
+	t.Log("claim created")
 
 	// wait until the controller pairs the volume and claim
-	waitForPersistentVolumePhase(w, api.VolumeBound)
+	waitForPersistentVolumePhase(watchPV, api.VolumeBound)
+	t.Log("volume bound")
+	waitForPersistentVolumeClaimPhase(watchPVC, api.ClaimBound)
+	t.Log("claim bound")
 
 	// only RWM PV is bound
 	pv, err := testClient.PersistentVolumes().Get("pv-rwo")
@@ -310,8 +310,10 @@ func TestPersistentVolumeMultiPVsDiffAccessModes(t *testing.T) {
 	if err := testClient.PersistentVolumeClaims(api.NamespaceDefault).Delete(pvc.Name, nil); err != nil {
 		t.Errorf("error deleting claim %s", pvc.Name)
 	}
+	t.Log("claim deleted")
 
-	waitForPersistentVolumePhase(w, api.VolumeReleased)
+	waitForPersistentVolumePhase(watchPV, api.VolumeReleased)
+	t.Log("volume released")
 
 	deleteAllEtcdKeys()
 }
@@ -326,7 +328,17 @@ func waitForPersistentVolumePhase(w watch.Interface, phase api.PersistentVolumeP
 	}
 }
 
-func createClients(s *httptest.Server) (*clientset.Clientset, *persistentvolumecontroller.PersistentVolumeController) {
+func waitForPersistentVolumeClaimPhase(w watch.Interface, phase api.PersistentVolumeClaimPhase) {
+	for {
+		event := <-w.ResultChan()
+		claim := event.Object.(*api.PersistentVolumeClaim)
+		if claim.Status.Phase == phase {
+			break
+		}
+	}
+}
+
+func createClients(t *testing.T, s *httptest.Server) (*clientset.Clientset, *persistentvolumecontroller.PersistentVolumeController, watch.Interface, watch.Interface) {
 	// Use higher QPS and Burst, there is a test for race condition below, which
 	// creates many claims and default values were too low.
 	testClient := clientset.NewForConfigOrDie(&restclient.Config{Host: s.URL, ContentConfig: restclient.ContentConfig{GroupVersion: testapi.Default.GroupVersion()}, QPS: 1000, Burst: 100000})
@@ -345,7 +357,17 @@ func createClients(s *httptest.Server) (*clientset.Clientset, *persistentvolumec
 	}}
 	cloud := &fake_cloud.FakeCloud{}
 	ctrl := persistentvolumecontroller.NewPersistentVolumeController(testClient, 10*time.Second, nil, plugins, cloud, "", nil, nil, nil)
-	return testClient, ctrl
+
+	watchPV, err := testClient.PersistentVolumes().Watch(api.ListOptions{})
+	if err != nil {
+		t.Fatalf("Failed to watch PersistentVolumes: %v", err)
+	}
+	watchPVC, err := testClient.PersistentVolumeClaims(api.NamespaceDefault).Watch(api.ListOptions{})
+	if err != nil {
+		t.Fatalf("Failed to watch PersistentVolumeClaimss: %v", err)
+	}
+
+	return testClient, ctrl, watchPV, watchPVC
 }
 
 func createPV(name, path, cap string, mode []api.PersistentVolumeAccessMode, reclaim api.PersistentVolumeReclaimPolicy) *api.PersistentVolume {

